<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_91) on Fri Apr 28 13:43:19 CEST 2017 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>GenomicElement</title>
<meta name="date" content="2017-04-28">
<link rel="stylesheet" type="text/css" href="../../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="GenomicElement";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":9,"i13":10,"i14":10,"i15":10,"i16":10,"i17":10,"i18":9,"i19":10,"i20":10,"i21":10,"i22":9,"i23":10,"i24":10,"i25":9,"i26":9,"i27":9,"i28":9,"i29":9,"i30":9,"i31":9,"i32":9,"i33":9,"i34":9,"i35":9,"i36":10,"i37":10,"i38":10,"i39":10,"i40":10,"i41":10,"i42":10,"i43":10,"i44":10,"i45":10,"i46":9,"i47":9,"i48":9,"i49":9,"i50":9,"i51":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GenomicElement.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementComparator.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" target="_top">Frames</a></li>
<li><a href="GenomicElement.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</div>
<h2 title="Class GenomicElement" class="title">Class GenomicElement</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable</dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">GenomicElement</span>
extends java.lang.Object
implements java.io.Serializable</pre>
<div class="block">Utility handling genomic region overlaps.
 GenomicElement.java 04 OCT 2016</div>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../../serialized-form.html#at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#annotation">annotation</a></span></code>
<div class="block">annotation</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#CHR">CHR</a></span></code>
<div class="block">contig of this GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#END">END</a></span></code>
<div class="block">end of this GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ID">ID</a></span></code>
<div class="block">id of this GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#INTEGRAL">INTEGRAL</a></span></code>
<div class="block">integral of measures over this GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#MASKED">MASKED</a></span></code>
<div class="block">indicator whether this GenomicElement is masked</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#MAX">MAX</a></span></code>
<div class="block">max of measures over this GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#MEASURE">MEASURE</a></span></code>
<div class="block">measure associated with this GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.Vector</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#misc">misc</a></span></code>
<div class="block">Vector of miscellaneous features for this GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#POSITION">POSITION</a></span></code>
<div class="block">position of this GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.lang.StringBuilder</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#sb">sb</a></span></code>
<div class="block">StringBuilder</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#START">START</a></span></code>
<div class="block">start of this GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#STRAND">STRAND</a></span></code>
<div class="block">strand of this GenomicElement</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">GenomicElement</a></span>(java.lang.String&nbsp;line,
              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">GenomicElement</a></span>(java.lang.String&nbsp;line,
              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-int-int-">GenomicElement</a></span>(java.lang.String&nbsp;chr,
              int&nbsp;start,
              int&nbsp;end)</code>
<div class="block">Creates new GenomicElement.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-java.lang.String-java.lang.String-int-int-double-">GenomicElement</a></span>(java.lang.String&nbsp;id,
              java.lang.String&nbsp;chr,
              java.lang.String&nbsp;strand,
              int&nbsp;start,
              int&nbsp;end,
              double&nbsp;measure)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">GenomicElement</a></span>(java.lang.String&nbsp;chr,
              java.lang.String&nbsp;start,
              java.lang.String&nbsp;end,
              java.lang.String&nbsp;measure)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">GenomicElement</a></span>(java.lang.String&nbsp;id,
              java.lang.String&nbsp;chr,
              java.lang.String&nbsp;strand,
              java.lang.String&nbsp;start,
              java.lang.String&nbsp;end)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">GenomicElement</a></span>(java.lang.String&nbsp;id,
              java.lang.String&nbsp;chr,
              java.lang.String&nbsp;strand,
              java.lang.String&nbsp;start,
              java.lang.String&nbsp;end,
              java.lang.String&nbsp;measure)</code>
<div class="block">constructs a GenomicElement object</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#GenomicElement-htsjdk.variant.variantcontext.VariantContext-">GenomicElement</a></span>(htsjdk.variant.variantcontext.VariantContext&nbsp;vc)</code>
<div class="block">Creates new GenomicElement.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#AND-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">AND</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnDifferentStrands</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnSameStrand</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTB-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTB</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnDifferentStrands</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                       <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the different strands</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnSameStrand</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                 <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the same strand</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#deduplicate-java.util.Vector-">deduplicate</a></span>(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">deduplicates a VectorVector&#60;GenomicElement&#62; of GenomicElements
 removes elements with identical chr, start and end positions and leaves only one copy</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#get3primeCopy--">get3primeCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 3prime position of the input GenomicElement</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#get5primeCopy--">get5primeCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 5prime position of the input GenomicElement</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getAssembly--">getAssembly</a></span>()</code>
<div class="block">getter for assembly</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getChromosomeNumber--">getChromosomeNumber</a></span>()</code>
<div class="block">Returns chromosome number without leading "chr".</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getCoordinates--">getCoordinates</a></span>()</code>
<div class="block">Returns tab separated coordinates.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getCoordinates-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getCoordinates</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">returns coordinates in colon-minus form</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getEnd--">getEnd</a></span>()</code>
<div class="block">Returns end position.</div>
</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getEndCopy--">getEndCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the end position of the input GenomicElement</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getID--">getID</a></span>()</code>
<div class="block">getter for ID</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getMean--">getMean</a></span>()</code>
<div class="block">returns center position</div>
</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getMeanCopy--">getMeanCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the mean position of the input GenomicElement</div>
</td>
</tr>
<tr id="i18" class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getMiscContent-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getMiscContent</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">reads out miscellaneous information Vector</div>
</td>
</tr>
<tr id="i19" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getSb--">getSb</a></span>()</code>
<div class="block">getter for StringBuilder info</div>
</td>
</tr>
<tr id="i20" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getStart--">getStart</a></span>()</code>
<div class="block">Returns start position.</div>
</td>
</tr>
<tr id="i21" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getStartCopy--">getStartCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the start position of the input GenomicElement</div>
</td>
</tr>
<tr id="i22" class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getTSS-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getTSS</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">return 5'end</div>
</td>
</tr>
<tr id="i23" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getUCSCLink--">getUCSCLink</a></span>()</code>
<div class="block">getter for USCS hyperlink</div>
</td>
</tr>
<tr id="i24" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#initMisc--">initMisc</a></span>()</code>
<div class="block">init Vector for miscellaneous infos</div>
</td>
</tr>
<tr id="i25" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOT-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOT</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements</div>
</td>
</tr>
<tr id="i26" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnDifferentStrands</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr id="i27" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnSameStrand</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on same strand</div>
</td>
</tr>
<tr id="i28" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OR</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
  <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements</div>
</td>
</tr>
<tr id="i29" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnDifferentStrands</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                    <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr id="i30" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnSameStrand</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr id="i31" class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlaps-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlaps</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
        <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">tests if two GenomicElements overlap</div>
</td>
</tr>
<tr id="i32" class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapSize-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapSize</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
           <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">calculates overlap size of two GenomicElements</div>
</td>
</tr>
<tr id="i33" class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapsOrTouches-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapsOrTouches</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                 <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">tests if two GenomicElements overlap or touch each other</div>
</td>
</tr>
<tr id="i34" class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapsPresent-java.util.Vector-">overlapsPresent</a></span>(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">tests if Vector&#60;GenomicElement&#62; contains overlapping GenomicElements</div>
</td>
</tr>
<tr id="i35" class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapType-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapType</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
           <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">calculates type of overlap
 s1 = start element 1
 e1 = end element 1
 s2 = start element 2
 e2 = end element 2
 note that most cases listed here are unrealistic but the cases are exhaustive
 including cases of zero size genomic elements and elements with negative size 
 where end is smaller than start.</div>
</td>
</tr>
<tr id="i36" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseColonCoordinates-java.lang.String-">parseColonCoordinates</a></span>(java.lang.String&nbsp;s)</code>
<div class="block">parseColonCoordinates</div>
</td>
</tr>
<tr id="i37" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseTabDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">parseTabDelimitedCoordinates</a></span>(java.lang.String&nbsp;s,
                            <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</code>
<div class="block">parseTabDelimitedCoordinates</div>
</td>
</tr>
<tr id="i38" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseTabDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">parseTabDelimitedCoordinates</a></span>(java.lang.String&nbsp;s,
                            <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
                            <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</code>
<div class="block">parseTabDelimitedCoordinates</div>
</td>
</tr>
<tr id="i39" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseWhiteSpaceDelimitedCoordinates-java.lang.String:A-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">parseWhiteSpaceDelimitedCoordinates</a></span>(java.lang.String[]&nbsp;sa,
                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</code>
<div class="block">parseWhiteSpaceDelimitedCoordinates</div>
</td>
</tr>
<tr id="i40" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseWhiteSpaceDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">parseWhiteSpaceDelimitedCoordinates</a></span>(java.lang.String&nbsp;s,
                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</code>
<div class="block">parseWhiteSpaceDelimitedCoordinates</div>
</td>
</tr>
<tr id="i41" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#parseWhiteSpaceDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">parseWhiteSpaceDelimitedCoordinates</a></span>(java.lang.String&nbsp;s,
                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</code>
<div class="block">parseWhiteSpaceDelimitedCoordinates.</div>
</td>
</tr>
<tr id="i42" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#setAssembly-java.lang.String-">setAssembly</a></span>(java.lang.String&nbsp;assembly)</code>
<div class="block">setter for assembly</div>
</td>
</tr>
<tr id="i43" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#setID-java.lang.String-">setID</a></span>(java.lang.String&nbsp;ID)</code>
<div class="block">setter for ID</div>
</td>
</tr>
<tr id="i44" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#setSb-java.lang.StringBuilder-">setSb</a></span>(java.lang.StringBuilder&nbsp;sb)</code>
<div class="block">setter for StringBuilder</div>
</td>
</tr>
<tr id="i45" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#setUCSCLink-java.lang.String-">setUCSCLink</a></span>(java.lang.String&nbsp;UCSCLink)</code>
<div class="block">setter for USCS hyperlink</div>
</td>
</tr>
<tr id="i46" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#setWarningCounter-int-">setWarningCounter</a></span>(int&nbsp;count)</code>
<div class="block">Sets the warning counter.</div>
</td>
</tr>
<tr id="i47" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#shrinkToNonOverlapping-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">shrinkToNonOverlapping</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                      <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">shrinks two overlapping GenomicElements to two non-overlapping GenomicElements</div>
</td>
</tr>
<tr id="i48" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOR</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">Returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements</div>
</td>
</tr>
<tr id="i49" class="rowColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-java.util.Vector-">XOR</a></span>(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of the overlapping GenomicElements</div>
</td>
</tr>
<tr id="i50" class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnDifferentStrands</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr id="i51" class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnSameStrand</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="ID">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ID</h4>
<pre>public&nbsp;java.lang.String ID</pre>
<div class="block">id of this GenomicElement</div>
</li>
</ul>
<a name="CHR">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CHR</h4>
<pre>public&nbsp;java.lang.String CHR</pre>
<div class="block">contig of this GenomicElement</div>
</li>
</ul>
<a name="STRAND">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>STRAND</h4>
<pre>public&nbsp;java.lang.String STRAND</pre>
<div class="block">strand of this GenomicElement</div>
</li>
</ul>
<a name="START">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>START</h4>
<pre>public&nbsp;int START</pre>
<div class="block">start of this GenomicElement</div>
</li>
</ul>
<a name="END">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>END</h4>
<pre>public&nbsp;int END</pre>
<div class="block">end of this GenomicElement</div>
</li>
</ul>
<a name="MEASURE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MEASURE</h4>
<pre>public&nbsp;double MEASURE</pre>
<div class="block">measure associated with this GenomicElement</div>
</li>
</ul>
<a name="INTEGRAL">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>INTEGRAL</h4>
<pre>public&nbsp;double INTEGRAL</pre>
<div class="block">integral of measures over this GenomicElement</div>
</li>
</ul>
<a name="MAX">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MAX</h4>
<pre>public&nbsp;double MAX</pre>
<div class="block">max of measures over this GenomicElement</div>
</li>
</ul>
<a name="POSITION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>POSITION</h4>
<pre>public&nbsp;int POSITION</pre>
<div class="block">position of this GenomicElement</div>
</li>
</ul>
<a name="misc">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>misc</h4>
<pre>public&nbsp;java.util.Vector misc</pre>
<div class="block">Vector of miscellaneous features for this GenomicElement</div>
</li>
</ul>
<a name="sb">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>sb</h4>
<pre>public&nbsp;java.lang.StringBuilder sb</pre>
<div class="block">StringBuilder</div>
</li>
</ul>
<a name="annotation">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>annotation</h4>
<pre>public&nbsp;java.lang.String annotation</pre>
<div class="block">annotation</div>
</li>
</ul>
<a name="MASKED">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MASKED</h4>
<pre>public&nbsp;boolean MASKED</pre>
<div class="block">indicator whether this GenomicElement is masked</div>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="GenomicElement-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;line,
                      <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>line</code> - a line of the bed file</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;line,
                      <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
                      <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>line</code> - a line of the bed file</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
<dd><code>gui</code> - graphical user interface</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;id,
                      java.lang.String&nbsp;chr,
                      java.lang.String&nbsp;strand,
                      java.lang.String&nbsp;start,
                      java.lang.String&nbsp;end)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>id</code> - coordinates</dd>
<dd><code>chr</code> - chr</dd>
<dd><code>strand</code> - strand</dd>
<dd><code>start</code> - start</dd>
<dd><code>end</code> - end</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-int-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;chr,
                      int&nbsp;start,
                      int&nbsp;end)</pre>
<div class="block">Creates new GenomicElement.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>chr</code> - chr</dd>
<dd><code>start</code> - start</dd>
<dd><code>end</code> - end</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-htsjdk.variant.variantcontext.VariantContext-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(htsjdk.variant.variantcontext.VariantContext&nbsp;vc)</pre>
<div class="block">Creates new GenomicElement.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>vc</code> - VariantContext object</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;id,
                      java.lang.String&nbsp;chr,
                      java.lang.String&nbsp;strand,
                      java.lang.String&nbsp;start,
                      java.lang.String&nbsp;end,
                      java.lang.String&nbsp;measure)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>id</code> - coordiantes</dd>
<dd><code>chr</code> - chr</dd>
<dd><code>strand</code> - strand</dd>
<dd><code>start</code> - start</dd>
<dd><code>end</code> - end</dd>
<dd><code>measure</code> - measure associated with this region</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-java.lang.String-java.lang.String-int-int-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;id,
                      java.lang.String&nbsp;chr,
                      java.lang.String&nbsp;strand,
                      int&nbsp;start,
                      int&nbsp;end,
                      double&nbsp;measure)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>id</code> - coordinates</dd>
<dd><code>chr</code> - chr</dd>
<dd><code>strand</code> - strand</dd>
<dd><code>start</code> - start</dd>
<dd><code>end</code> - end</dd>
<dd><code>measure</code> - measure associated with this region</dd>
</dl>
</li>
</ul>
<a name="GenomicElement-java.lang.String-java.lang.String-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>GenomicElement</h4>
<pre>public&nbsp;GenomicElement(java.lang.String&nbsp;chr,
                      java.lang.String&nbsp;start,
                      java.lang.String&nbsp;end,
                      java.lang.String&nbsp;measure)</pre>
<div class="block">constructs a GenomicElement object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>chr</code> - chr</dd>
<dd><code>start</code> - start</dd>
<dd><code>end</code> - end</dd>
<dd><code>measure</code> - measure associated with this region</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getTSS-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTSS</h4>
<pre>public static&nbsp;int&nbsp;getTSS(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</pre>
<div class="block">return 5'end</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int</dd>
</dl>
</li>
</ul>
<a name="getMean--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMean</h4>
<pre>public&nbsp;int&nbsp;getMean()</pre>
<div class="block">returns center position</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int</dd>
</dl>
</li>
</ul>
<a name="getStartCopy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getStartCopy</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;getStartCopy()</pre>
<div class="block">returns a GenomicElement object with length 1 corresponding to the start position of the input GenomicElement</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="getEndCopy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getEndCopy</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;getEndCopy()</pre>
<div class="block">returns a GenomicElement object with length 1 corresponding to the end position of the input GenomicElement</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="getMeanCopy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanCopy</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;getMeanCopy()</pre>
<div class="block">returns a GenomicElement object with length 1 corresponding to the mean position of the input GenomicElement</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="get5primeCopy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>get5primeCopy</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;get5primeCopy()</pre>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 5prime position of the input GenomicElement</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="get3primeCopy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>get3primeCopy</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;get3primeCopy()</pre>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 3prime position of the input GenomicElement</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="initMisc--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initMisc</h4>
<pre>public&nbsp;void&nbsp;initMisc()</pre>
<div class="block">init Vector for miscellaneous infos</div>
</li>
</ul>
<a name="getMiscContent-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMiscContent</h4>
<pre>public static&nbsp;java.lang.String&nbsp;getMiscContent(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</pre>
<div class="block">reads out miscellaneous information Vector</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
</dl>
</li>
</ul>
<a name="overlapSize-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapSize</h4>
<pre>public static&nbsp;int&nbsp;overlapSize(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">calculates overlap size of two GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int</dd>
</dl>
</li>
</ul>
<a name="overlaps-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlaps</h4>
<pre>public static&nbsp;boolean&nbsp;overlaps(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                               <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">tests if two GenomicElements overlap</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>boolean</dd>
</dl>
</li>
</ul>
<a name="overlapsOrTouches-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsOrTouches</h4>
<pre>public static&nbsp;boolean&nbsp;overlapsOrTouches(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                        <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">tests if two GenomicElements overlap or touch each other</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>boolean</dd>
</dl>
</li>
</ul>
<a name="shrinkToNonOverlapping-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>shrinkToNonOverlapping</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;shrinkToNonOverlapping(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                      <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">shrinks two overlapping GenomicElements to two non-overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement[]</dd>
</dl>
</li>
</ul>
<a name="OR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>OR</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;OR(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="OROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>OROnSameStrand</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;OROnSameStrand(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                            <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on the same strand</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="OROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>OROnDifferentStrands</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;OROnDifferentStrands(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                  <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on different strands</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="AND-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>AND</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;AND(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                 <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="ANDOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ANDOnSameStrand</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ANDOnSameStrand(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                             <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on the same strand</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="ANDOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ANDOnDifferentStrands</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ANDOnDifferentStrands(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on different strands</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="deduplicate-java.util.Vector-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>deduplicate</h4>
<pre>public static&nbsp;java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;deduplicate(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</pre>
<div class="block">deduplicates a VectorVector&#60;GenomicElement&#62; of GenomicElements
 removes elements with identical chr, start and end positions and leaves only one copy</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>v</code> - the Vector object</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Vector&#60;GenomicElement&#62;</dd>
</dl>
</li>
</ul>
<a name="overlapsPresent-java.util.Vector-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapsPresent</h4>
<pre>public static&nbsp;boolean&nbsp;overlapsPresent(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</pre>
<div class="block">tests if Vector&#60;GenomicElement&#62; contains overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>v</code> - the Vector object</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>boolean</dd>
</dl>
</li>
</ul>
<a name="XOR-java.util.Vector-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>XOR</h4>
<pre>public static&nbsp;java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;XOR(java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</pre>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of the overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>v</code> - the Vector object</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Vector&#60;GenomicElement&#62;</dd>
</dl>
</li>
</ul>
<a name="XOR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>XOR</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;XOR(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">Returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement[]</dd>
</dl>
</li>
</ul>
<a name="XOROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>XOROnSameStrand</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;XOROnSameStrand(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                               <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on the same strand</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement[]</dd>
</dl>
</li>
</ul>
<a name="XOROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>XOROnDifferentStrands</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;XOROnDifferentStrands(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on different strands</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement[]</dd>
</dl>
</li>
</ul>
<a name="NOT-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NOT</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;NOT(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                 <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="NOTOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NOTOnSameStrand</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;NOTOnSameStrand(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                             <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on same strand</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="NOTOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NOTOnDifferentStrands</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;NOTOnDifferentStrands(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                   <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on different strands</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="ANOTB-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ANOTB</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;ANOTB(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="ANOTBOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ANOTBOnSameStrand</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;ANOTBOnSameStrand(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                 <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the same strand</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="ANOTBOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ANOTBOnDifferentStrands</h4>
<pre>public static&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;ANOTBOnDifferentStrands(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                                                       <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the different strands</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - genomic element</dd>
<dd><code>ge2</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>GenomicElement</dd>
</dl>
</li>
</ul>
<a name="overlapType-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>overlapType</h4>
<pre>public static&nbsp;int&nbsp;overlapType(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                              <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</pre>
<div class="block">calculates type of overlap
 s1 = start element 1
 e1 = end element 1
 s2 = start element 2
 e2 = end element 2
 note that most cases listed here are unrealistic but the cases are exhaustive
 including cases of zero size genomic elements and elements with negative size 
 where end is smaller than start.
 
 In total there are 81 cases to consider: 
 comparisons to be made are between s1 vs s2, s1 vs e2, e1 vs s2, and e1 vs e2
 each comparison can be &#60;, =, or &#62;.
 Thus the total number of cases is 3^4 = 81.

case  1 s1&#60;s2       e1&#60;s2       s1&#60;e2       e1&#60;e2
case  2 s1&#60;s2       e1&#60;s2       s1&#60;e2       e1=e2
case  3 s1&#60;s2       e1&#60;s2       s1&#60;e2       e1&#62;e2
case  4 s1&#60;s2       e1&#60;s2       s1=e2   e1&#60;e2
case  5 s1&#60;s2       e1&#60;s2       s1=e2   e1=e2
case  6 s1&#60;s2       e1&#60;s2       s1=e2   e1&#62;e2
case  7 s1&#60;s2       e1&#60;s2       s1&#62;e2       e1&#60;e2
case  8 s1&#60;s2       e1&#60;s2       s1&#62;e2       e1=e2
case  9 s1&#60;s2       e1&#60;s2       s1&#62;e2       e1&#62;e2
case 10 s1&#60;s2       e1=s2   s1&#60;e2       e1&#60;e2
case 11 s1&#60;s2       e1=s2   s1&#60;e2       e1=e2
case 12 s1&#60;s2       e1=s2   s1&#60;e2       e1&#62;e2
case 13 s1&#60;s2       e1=s2   s1=e2   e1&#60;e2
case 14 s1&#60;s2       e1=s2   s1=e2   e1=e2
case 15 s1&#60;s2       e1=s2   s1=e2   e1&#62;e2
case 16 s1&#60;s2       e1=s2   s1&#62;e2       e1&#60;e2
case 17 s1&#60;s2       e1=s2   s1&#62;e2       e1=e2
case 18 s1&#60;s2       e1=s2   s1&#62;e2       e1&#62;e2
case 19 s1&#60;s2       e1&#62;s2       s1&#60;e2       e1&#60;e2
case 20 s1&#60;s2       e1&#62;s2       s1&#60;e2       e1=e2
case 21 s1&#60;s2       e1&#62;s2       s1&#60;e2       e1&#62;e2
case 22 s1&#60;s2       e1&#62;s2       s1=e2   e1&#60;e2
case 23 s1&#60;s2       e1&#62;s2       s1=e2   e1=e2
case 24 s1&#60;s2       e1&#62;s2       s1=e2   e1&#62;e2
case 25 s1&#60;s2       e1&#62;s2       s1&#62;e2       e1&#60;e2
case 26 s1&#60;s2       e1&#62;s2       s1&#62;e2       e1=e2
case 27 s1&#60;s2       e1&#62;s2       s1&#62;e2       e1&#62;e2
case 28 s1=s2   e1&#60;s2       s1&#60;e2       e1&#60;e2
case 29 s1=s2   e1&#60;s2       s1&#60;e2       e1=e2
case 30 s1=s2   e1&#60;s2       s1&#60;e2       e1&#62;e2
case 31 s1=s2   e1&#60;s2       s1=e2   e1&#60;e2
case 32 s1=s2   e1&#60;s2       s1=e2   e1=e2
case 33 s1=s2   e1&#60;s2       s1=e2   e1&#62;e2
case 34 s1=s2   e1&#60;s2       s1&#62;e2       e1&#60;e2
case 35 s1=s2   e1&#60;s2       s1&#62;e2       e1=e2
case 36 s1=s2   e1&#60;s2       s1&#62;e2       e1&#62;e2
case 37 s1=s2   e1=s2   s1&#60;e2       e1&#60;e2
case 38 s1=s2   e1=s2   s1&#60;e2       e1=e2
case 39 s1=s2   e1=s2   s1&#60;e2       e1&#62;e2
case 40 s1=s2   e1=s2   s1=e2   e1&#60;e2
case 41 s1=s2   e1=s2   s1=e2   e1=e2
case 42 s1=s2   e1=s2   s1=e2   e1&#62;e2
case 43 s1=s2   e1=s2   s1&#62;e2       e1&#60;e2
case 44 s1=s2   e1=s2   s1&#62;e2       e1=e2
case 45 s1=s2   e1=s2   s1&#62;e2       e1&#62;e2
case 46 s1=s2   e1&#62;s2       s1&#60;e2       e1&#60;e2
case 47 s1=s2   e1&#62;s2       s1&#60;e2       e1=e2
case 48 s1=s2   e1&#62;s2       s1&#60;e2       e1&#62;e2
case 49 s1=s2   e1&#62;s2       s1=e2   e1&#60;e2
case 50 s1=s2   e1&#62;s2       s1=e2   e1=e2
case 51 s1=s2   e1&#62;s2       s1=e2   e1&#62;e2
case 52 s1=s2   e1&#62;s2       s1&#62;e2       e1&#60;e2
case 53 s1=s2   e1&#62;s2       s1&#62;e2       e1=e2
case 54 s1=s2   e1&#62;s2       s1&#62;e2       e1&#62;e2
case 55 s1&#62;s2       e1&#60;s2       s1&#60;e2       e1&#60;e2
case 56 s1&#62;s2       e1&#60;s2       s1&#60;e2       e1=e2
case 57 s1&#62;s2       e1&#60;s2       s1&#60;e2       e1&#62;e2
case 58 s1&#62;s2       e1&#60;s2       s1=e2   e1&#60;e2
case 59 s1&#62;s2       e1&#60;s2       s1=e2   e1=e2
case 60 s1&#62;s2       e1&#60;s2       s1=e2   e1&#62;e2
case 61 s1&#62;s2       e1&#60;s2       s1&#62;e2       e1&#60;e2
case 62 s1&#62;s2       e1&#60;s2       s1&#62;e2       e1=e2
case 63 s1&#62;s2       e1&#60;s2       s1&#62;e2       e1&#62;e2
case 64 s1&#62;s2       e1=s2   s1&#60;e2       e1&#60;e2
case 65 s1&#62;s2       e1=s2   s1&#60;e2       e1=e2
case 66 s1&#62;s2       e1=s2   s1&#60;e2       e1&#62;e2
case 67 s1&#62;s2       e1=s2   s1=e2   e1&#60;e2
case 68 s1&#62;s2       e1=s2   s1=e2   e1=e2
case 69 s1&#62;s2       e1=s2   s1=e2   e1&#62;e2
case 70 s1&#62;s2       e1=s2   s1&#62;e2       e1&#60;e2
case 71 s1&#62;s2       e1=s2   s1&#62;e2       e1=e2
case 72 s1&#62;s2       e1=s2   s1&#62;e2       e1&#62;e2
case 73 s1&#62;s2       e1&#62;s2       s1&#60;e2       e1&#60;e2
case 74 s1&#62;s2       e1&#62;s2       s1&#60;e2       e1=e2
case 75 s1&#62;s2       e1&#62;s2       s1&#60;e2       e1&#62;e2
case 76 s1&#62;s2       e1&#62;s2       s1=e2   e1&#60;e2
case 77 s1&#62;s2       e1&#62;s2       s1=e2   e1=e2
case 78 s1&#62;s2       e1&#62;s2       s1=e2   e1&#62;e2
case 79 s1&#62;s2       e1&#62;s2       s1&#62;e2       e1&#60;e2
case 80 s1&#62;s2       e1&#62;s2       s1&#62;e2       e1=e2
case 81 s1&#62;s2       e1&#62;s2       s1&#62;e2       e1&#62;e2

 correspondence relationships for realistic cases
 when overlapType(GenomicElement e) is called for element1 with element2 as parameter or vice versa

 cases tested:

 GenomicElement ge = new GenomicElement("ID1", "chr1", "+", "100", "1000");

   GenomicElement ge1 = new GenomicElement("ID1", "chr1", "+", "10000", "1000000");
   GenomicElement ge2 = new GenomicElement("ID1", "chr1", "+", "1000", "100000");
   GenomicElement ge3 = new GenomicElement("ID1", "chr1", "+", "500", "100000");
   GenomicElement ge4 = new GenomicElement("ID1", "chr1", "+", "100", "100000");
   GenomicElement ge5 = new GenomicElement("ID1", "chr1", "+", "50", "100000");
   GenomicElement ge6 = new GenomicElement("ID1", "chr1", "+", "200", "1000");
   GenomicElement ge7 = new GenomicElement("ID1", "chr1", "+", "200", "900");
   GenomicElement ge8 = new GenomicElement("ID1", "chr1", "+", "100", "1000");
   GenomicElement ge9 = new GenomicElement("ID1", "chr1", "+", "100", "900");
   GenomicElement ge10 = new GenomicElement("ID1", "chr1", "+", "50", "1000");
   GenomicElement ge11 = new GenomicElement("ID1", "chr1", "+", "50", "500");
   GenomicElement ge12 = new GenomicElement("ID1", "chr1", "+", "50", "100");
   GenomicElement ge13 = new GenomicElement("ID1", "chr1", "+", "50", "60");

        case element1   case element 2
1       1               81              element 1 upstream of element2, no overlap, no touch
2       10              78              element 1 upstream of element2, no overlap, touch
3       19              75              s1&#60;s2       e1&#62;s2       s1&#60;e2       e1&#60;e2 partial overlap with element 1 upstream
4       46              48              s1=s2   e1&#62;s2       s1&#60;e2       e1&#60;e2 element 1 contained in element2 starts identical
5       73              21              s1&#62;s2       e1&#62;s2       s1&#60;e2       e1&#60;e2 element 1 contained in element2
6       20              74              s1&#60;s2       e1&#62;s2       s1&#60;e2       e1=e2 element 2 contained in element1 ends identical
7       21              73              s1&#60;s2       e1&#62;s2       s1&#60;e2       e1&#62;e2 element 2 contained in element1
8       47              47              elements are identical and size is &#62; 0
9       48              46              s1=s2   e1&#62;s2       s1&#60;e2       e1&#62;e2 element 2 contained in element1 starts identical
10      74              20              s1&#62;s2       e1&#62;s2       s1&#60;e2       e1=e2 element 1 contained in element2 ends identical
11      75              19              s1&#62;s2       e1&#62;s2       s1&#60;e2       e1&#62;e2 partial overlap with element 2 upstream
12      78              10              element 2 upstream of element1, no overlap, touch
13      81              1               element 2 upstream of element1, no overlap, no touch</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge1</code> - GenomicElement to be tested</dd>
<dd><code>ge2</code> - GenomicElement to be tested</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int code of overlap type</dd>
</dl>
</li>
</ul>
<a name="getCoordinates-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getCoordinates</h4>
<pre>public static&nbsp;java.lang.String&nbsp;getCoordinates(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</pre>
<div class="block">returns coordinates in colon-minus form</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ge</code> - genomic element</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
</dl>
</li>
</ul>
<a name="parseColonCoordinates-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseColonCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseColonCoordinates(java.lang.String&nbsp;s)</pre>
<div class="block">parseColonCoordinates</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>s</code> - string of genomic coordinates</dd>
</dl>
</li>
</ul>
<a name="parseWhiteSpaceDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseWhiteSpaceDelimitedCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseWhiteSpaceDelimitedCoordinates(java.lang.String&nbsp;s,
                                                <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</pre>
<div class="block">parseWhiteSpaceDelimitedCoordinates</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>s</code> - string of genomic coordinates</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
</dl>
</li>
</ul>
<a name="parseWhiteSpaceDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseWhiteSpaceDelimitedCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseWhiteSpaceDelimitedCoordinates(java.lang.String&nbsp;s,
                                                <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
                                                <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</pre>
<div class="block">parseWhiteSpaceDelimitedCoordinates. Chromosome names are converted to upper case.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>s</code> - string of genomic coordinates</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
<dd><code>gui</code> - graphical user interface</dd>
</dl>
</li>
</ul>
<a name="parseTabDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseTabDelimitedCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseTabDelimitedCoordinates(java.lang.String&nbsp;s,
                                         <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</pre>
<div class="block">parseTabDelimitedCoordinates</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>s</code> - string of genomic coordinates</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
</dl>
</li>
</ul>
<a name="parseTabDelimitedCoordinates-java.lang.String-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseTabDelimitedCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseTabDelimitedCoordinates(java.lang.String&nbsp;s,
                                         <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr,
                                         <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</pre>
<div class="block">parseTabDelimitedCoordinates</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>s</code> - string of genomic coordinates</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
<dd><code>gui</code> - graphical user interface</dd>
</dl>
</li>
</ul>
<a name="parseWhiteSpaceDelimitedCoordinates-java.lang.String:A-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.FileTypeReader-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseWhiteSpaceDelimitedCoordinates</h4>
<pre>public&nbsp;void&nbsp;parseWhiteSpaceDelimitedCoordinates(java.lang.String[]&nbsp;sa,
                                                <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">FileTypeReader</a>&nbsp;ftr)</pre>
<div class="block">parseWhiteSpaceDelimitedCoordinates</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sa</code> - array of strings of genomic coordinates</dd>
<dd><code>ftr</code> - FileTypeReader object</dd>
</dl>
</li>
</ul>
<a name="getID--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getID</h4>
<pre>public&nbsp;java.lang.String&nbsp;getID()</pre>
<div class="block">getter for ID</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>return genomic coordinates</dd>
</dl>
</li>
</ul>
<a name="setID-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setID</h4>
<pre>public&nbsp;void&nbsp;setID(java.lang.String&nbsp;ID)</pre>
<div class="block">setter for ID</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>ID</code> - sets genomic coordinates</dd>
</dl>
</li>
</ul>
<a name="getSb--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSb</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSb()</pre>
<div class="block">getter for StringBuilder info</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
</dl>
</li>
</ul>
<a name="setSb-java.lang.StringBuilder-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSb</h4>
<pre>public&nbsp;void&nbsp;setSb(java.lang.StringBuilder&nbsp;sb)</pre>
<div class="block">setter for StringBuilder</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sb</code> - sets StringBuilder</dd>
</dl>
</li>
</ul>
<a name="getUCSCLink--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getUCSCLink</h4>
<pre>public&nbsp;java.lang.String&nbsp;getUCSCLink()</pre>
<div class="block">getter for USCS hyperlink</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
</dl>
</li>
</ul>
<a name="setUCSCLink-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setUCSCLink</h4>
<pre>public&nbsp;void&nbsp;setUCSCLink(java.lang.String&nbsp;UCSCLink)</pre>
<div class="block">setter for USCS hyperlink</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>UCSCLink</code> - sets UCSC hyperlink</dd>
</dl>
</li>
</ul>
<a name="getAssembly--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAssembly</h4>
<pre>public&nbsp;java.lang.String&nbsp;getAssembly()</pre>
<div class="block">getter for assembly</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
</dl>
</li>
</ul>
<a name="setAssembly-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setAssembly</h4>
<pre>public&nbsp;void&nbsp;setAssembly(java.lang.String&nbsp;assembly)</pre>
<div class="block">setter for assembly</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>assembly</code> - UCSCLink</dd>
</dl>
</li>
</ul>
<a name="getCoordinates--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getCoordinates</h4>
<pre>public&nbsp;java.lang.String&nbsp;getCoordinates()</pre>
<div class="block">Returns tab separated coordinates.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="getChromosomeNumber--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getChromosomeNumber</h4>
<pre>public&nbsp;java.lang.String&nbsp;getChromosomeNumber()</pre>
<div class="block">Returns chromosome number without leading "chr".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>String</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="getStart--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getStart</h4>
<pre>public&nbsp;int&nbsp;getStart()</pre>
<div class="block">Returns start position.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="getEnd--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getEnd</h4>
<pre>public&nbsp;int&nbsp;getEnd()</pre>
<div class="block">Returns end position.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>int</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="setWarningCounter-int-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>setWarningCounter</h4>
<pre>public static&nbsp;void&nbsp;setWarningCounter(int&nbsp;count)</pre>
<div class="block">Sets the warning counter.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>count</code> - max counts for warnings</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GenomicElement.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/FileTypeReader.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementComparator.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" target="_top">Frames</a></li>
<li><a href="GenomicElement.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
