<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_91) on Fri Apr 28 13:43:19 CEST 2017 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Query</title>
<meta name="date" content="2017-04-28">
<link rel="stylesheet" type="text/css" href="../../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Query";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Query.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementSort.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Sort.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html" target="_top">Frames</a></li>
<li><a href="Query.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</div>
<h2 title="Class Query" class="title">Class Query</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.Query</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable</dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">Query</span>
extends java.lang.Object
implements java.io.Serializable</pre>
<div class="block">Permits fast searching of overlaps between GenomicElements.
 Query.java 04 OCT 2016</div>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../../serialized-form.html#at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.Query">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#Query--">Query</a></span>()</code>
<div class="block">Creates an instance of RangeQuery.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#Query-java.util.ArrayList-">Query</a></span>(java.util.ArrayList&lt;java.io.File&gt;&nbsp;bedfiles)</code>
<div class="block">Creates new Query.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#Query-java.util.ArrayList-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">Query</a></span>(java.util.ArrayList&lt;java.io.File&gt;&nbsp;bedfiles,
     <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</code>
<div class="block">Creates new Query.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#Query-java.io.File:A-">Query</a></span>(java.io.File[]&nbsp;bedfiles)</code>
<div class="block">Creates new Query.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#complement-java.util.ArrayList-">complement</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#flatten-java.util.ArrayList-">flatten</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>java.util.Hashtable&lt;java.lang.String,java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getAnnotations--">getAnnotations</a></span>()</code>
<div class="block">getter for Regions Hashtable</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getFlattenedRegions--">getFlattenedRegions</a></span>()</code>
<div class="block">Returns flattened regions.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getFlattenedRegionsComplement--">getFlattenedRegionsComplement</a></span>()</code>
<div class="block">Returns flattened regions complement.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#initAnnotations-java.util.ArrayList-">initAnnotations</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in)</code>
<div class="block">private method that inits the Regions and Indices Hashtables.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#initPointAnnotations-java.util.ArrayList-java.lang.String-">initPointAnnotations</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in,
                    java.lang.String&nbsp;point)</code>
<div class="block">private method that inits the Regions and Indices Hashtables.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#match-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">match</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Tests for matching regions in Regions.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsQuery</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsStartQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsStartQuery</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#rangeQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">rangeQuery</a></span>(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="Query--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Query</h4>
<pre>public&nbsp;Query()</pre>
<div class="block">Creates an instance of RangeQuery.</div>
</li>
</ul>
<a name="Query-java.io.File:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Query</h4>
<pre>public&nbsp;Query(java.io.File[]&nbsp;bedfiles)</pre>
<div class="block">Creates new Query.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>bedfiles</code> - array of bed file filenames</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="Query-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Query</h4>
<pre>public&nbsp;Query(java.util.ArrayList&lt;java.io.File&gt;&nbsp;bedfiles)</pre>
<div class="block">Creates new Query.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>bedfiles</code> - list of bed file filenames</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="Query-java.util.ArrayList-at.ac.oeaw.cemm.bsf.vcffilter.VCFFilter-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>Query</h4>
<pre>public&nbsp;Query(java.util.ArrayList&lt;java.io.File&gt;&nbsp;bedfiles,
             <a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/VCFFilter.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter">VCFFilter</a>&nbsp;gui)</pre>
<div class="block">Creates new Query.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>bedfiles</code> - list of bed file filenames</dd>
<dd><code>gui</code> - graphical user interface</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="flatten-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>flatten</h4>
<pre>public&nbsp;java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;flatten(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</pre>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>regions</code> - list of genomic regions</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>ArrayList&#60;GenomicElement&#60;</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="complement-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>complement</h4>
<pre>public&nbsp;java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;complement(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</pre>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions. 
 Then complements the list.
 
 Currently not working correctly as contigs with 0 input regions are missing in output. 
 Proper implementation should read the contigs from the source VCF file.
 Useful for black list filtering.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>regions</code> - list of genomic regions</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>ArrayList&#60;GenomicElement&#60;</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="initAnnotations-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initAnnotations</h4>
<pre>public&nbsp;void&nbsp;initAnnotations(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in)</pre>
<div class="block">private method that inits the Regions and Indices Hashtables.
 The region elements for each chromosome are added as GenomicElement objects
 to separate Vector objects that can be retrieved by chromosome name.
 The index of the last overlapping region element is stored in the Indices Hashtable.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>in</code> - the annotations as GenomicElement containing ArrayList</dd>
</dl>
</li>
</ul>
<a name="initPointAnnotations-java.util.ArrayList-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initPointAnnotations</h4>
<pre>public&nbsp;void&nbsp;initPointAnnotations(java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in,
                                 java.lang.String&nbsp;point)</pre>
<div class="block">private method that inits the Regions and Indices Hashtables.
 The region elements for each chromosome are added as GenomicElement objects
 to separate Vector objects that can be retrieved by chromosome name.
 The index of the last overlapping region element is stored in the Indices Hashtable.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>in</code> - the annotations as GenomicElement containing ArrayList</dd>
<dd><code>point</code> - start point</dd>
</dl>
</li>
</ul>
<a name="rangeQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>rangeQuery</h4>
<pre>public&nbsp;java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;rangeQuery(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</pre>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>query</code> - defines the range being interrogated</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Vector holding all GenomicElements overlapping with query range, Vector size() == 0 if no overlap exists</dd>
</dl>
</li>
</ul>
<a name="match-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>match</h4>
<pre>public&nbsp;boolean&nbsp;match(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</pre>
<div class="block">Tests for matching regions in Regions.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>query</code> - the query region</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>boolean</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="neighborsQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>neighborsQuery</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;neighborsQuery(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</pre>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>query</code> - defines the range being interrogated</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Vector holding all GenomicElements overlapping with query range, Vector size() == 0 if no overlap exists</dd>
</dl>
</li>
</ul>
<a name="neighborsStartQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>neighborsStartQuery</h4>
<pre>public&nbsp;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]&nbsp;neighborsStartQuery(<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</pre>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>query</code> - defines the range being interrogated</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Vector holding all GenomicElements overlapping with query range, Vector size() == 0 if no overlap exists</dd>
</dl>
</li>
</ul>
<a name="getAnnotations--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAnnotations</h4>
<pre>public&nbsp;java.util.Hashtable&lt;java.lang.String,java.util.Vector&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&gt;&nbsp;getAnnotations()</pre>
<div class="block">getter for Regions Hashtable</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Hashtable&#60;String, Vector&#60;GenomicElement&#62;&#62;</dd>
</dl>
</li>
</ul>
<a name="getFlattenedRegions--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getFlattenedRegions</h4>
<pre>public&nbsp;java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;getFlattenedRegions()</pre>
<div class="block">Returns flattened regions.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>ArrayList&#60;GenomicElement&#62;</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
<a name="getFlattenedRegionsComplement--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getFlattenedRegionsComplement</h4>
<pre>public&nbsp;java.util.ArrayList&lt;<a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;getFlattenedRegionsComplement()</pre>
<div class="block">Returns flattened regions complement.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>ArrayList&#60;GenomicElement&#62;</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>1.0</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Query.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementSort.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Sort.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html" target="_top">Frames</a></li>
<li><a href="Query.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
