<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_91) on Fri Apr 28 13:43:28 CEST 2017 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Uses of Class at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement</title>
<meta name="date" content="2017-04-28">
<link rel="stylesheet" type="text/css" href="../../../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Class at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../../../../../../../../overview-tree.html">Tree</a></li>
<li><a href="../../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/class-use/GenomicElement.html" target="_top">Frames</a></li>
<li><a href="GenomicElement.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Class at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement" class="title">Uses of Class<br>at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a></td>
<td class="colLast">&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a> in <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/package-summary.html">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/package-summary.html">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a> that return <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#AND-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">AND</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTB-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTB</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                       <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the different strands</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                 <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the same strand</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#get3primeCopy--">get3primeCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 3prime position of the input GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#get5primeCopy--">get5primeCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the 5prime position of the input GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getEndCopy--">getEndCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the end position of the input GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getMeanCopy--">getMeanCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the mean position of the input GenomicElement</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getStartCopy--">getStartCopy</a></span>()</code>
<div class="block">returns a GenomicElement object with length 1 corresponding to the start position of the input GenomicElement</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsStartQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsStartQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOT-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOT</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on same strand</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OR</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
  <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                    <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
              <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#shrinkToNonOverlapping-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">shrinkToNonOverlapping</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                      <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">shrinks two overlapping GenomicElements to two non-overlapping GenomicElements</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOR</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">Returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/package-summary.html">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a> that return types with arguments of type <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#complement-java.util.ArrayList-">complement</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#deduplicate-java.util.Vector-">deduplicate</a></span>(java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">deduplicates a VectorVector&#60;GenomicElement&#62; of GenomicElements
 removes elements with identical chr, start and end positions and leaves only one copy</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#flatten-java.util.ArrayList-">flatten</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.Hashtable&lt;java.lang.String,java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getAnnotations--">getAnnotations</a></span>()</code>
<div class="block">getter for Regions Hashtable</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getFlattenedRegions--">getFlattenedRegions</a></span>()</code>
<div class="block">Returns flattened regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#getFlattenedRegionsComplement--">getFlattenedRegionsComplement</a></span>()</code>
<div class="block">Returns flattened regions complement.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElementListParser.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementListParser.html#getGenomicElements--">getGenomicElements</a></span>()</code>
<div class="block">Returns list of parsed GenomicElements.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#rangeQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">rangeQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElementSort.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementSort.html#sortByPosition-java.util.ArrayList-java.lang.String-">sortByPosition</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;ge,
              java.lang.String&nbsp;sortPosition)</code>
<div class="block">sortByPosition sort list of GenomicElements by chromosome and then by start or end or center</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-java.util.Vector-">XOR</a></span>(java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of the overlapping GenomicElements</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/package-summary.html">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a> with parameters of type <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#AND-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">AND</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANDOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANDOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the region covered by both of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTB-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTB</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                       <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the different strands</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#ANOTBOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">ANOTBOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                 <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by ge1 but not ge2 on the same strand</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElementComparator.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementComparator.html#compare-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">compare</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;g1,
       <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;g2)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getCoordinates-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getCoordinates</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">returns coordinates in colon-minus form</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getMiscContent-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getMiscContent</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">reads out miscellaneous information Vector</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#getTSS-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">getTSS</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge)</code>
<div class="block">return 5'end</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#match-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">match</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Tests for matching regions in Regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#neighborsStartQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">neighborsStartQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOT-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOT</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#NOTOnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">NOTOnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions not covered by one of two overlapping GenomicElements on same strand</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OR</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
  <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                    <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#OROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">OROnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
              <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by at least one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlaps-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlaps</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
        <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">tests if two GenomicElements overlap</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapSize-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapSize</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
           <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">calculates overlap size of two GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapsOrTouches-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapsOrTouches</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                 <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">tests if two GenomicElements overlap or touch each other</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapType-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">overlapType</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
           <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">calculates type of overlap
 s1 = start element 1
 e1 = end element 1
 s2 = start element 2
 e2 = end element 2
 note that most cases listed here are unrealistic but the cases are exhaustive
 including cases of zero size genomic elements and elements with negative size 
 where end is smaller than start.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#rangeQuery-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">rangeQuery</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;query)</code>
<div class="block">Returns the GenomicElements that overlap with the query interval stacked in a Vector object</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#shrinkToNonOverlapping-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">shrinkToNonOverlapping</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                      <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">shrinks two overlapping GenomicElements to two non-overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOR</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
   <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">Returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnDifferentStrands-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnDifferentStrands</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
                     <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on different strands</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>[]</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOROnSameStrand-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-at.ac.oeaw.cemm.bsf.vcffilter.genomeutils.GenomicElement-">XOROnSameStrand</a></span>(<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge1,
               <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&nbsp;ge2)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of two overlapping GenomicElements on the same strand</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Method parameters in <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/package-summary.html">at.ac.oeaw.cemm.bsf.vcffilter.genomeutils</a> with type arguments of type <a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#complement-java.util.ArrayList-">complement</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#deduplicate-java.util.Vector-">deduplicate</a></span>(java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">deduplicates a VectorVector&#60;GenomicElement&#62; of GenomicElements
 removes elements with identical chr, start and end positions and leaves only one copy</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#flatten-java.util.ArrayList-">flatten</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;regions)</code>
<div class="block">Takes a list of regions and generates a non-overlapping list of regions covered by any of the input regions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#initAnnotations-java.util.ArrayList-">initAnnotations</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in)</code>
<div class="block">private method that inits the Regions and Indices Hashtables.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">Query.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/Query.html#initPointAnnotations-java.util.ArrayList-java.lang.String-">initPointAnnotations</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;in,
                    java.lang.String&nbsp;point)</code>
<div class="block">private method that inits the Regions and Indices Hashtables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElementSort.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementSort.html#isSorted-java.util.ArrayList-">isSorted</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;ge)</code>
<div class="block">tests if list of GenomicElements is sorted by chr and start position</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#overlapsPresent-java.util.Vector-">overlapsPresent</a></span>(java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">tests if Vector&#60;GenomicElement&#62; contains overlapping GenomicElements</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElementSort.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElementSort.html#sortByPosition-java.util.ArrayList-java.lang.String-">sortByPosition</a></span>(java.util.ArrayList&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;ge,
              java.lang.String&nbsp;sortPosition)</code>
<div class="block">sortByPosition sort list of GenomicElements by chromosome and then by start or end or center</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">GenomicElement.</span><code><span class="memberNameLink"><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html#XOR-java.util.Vector-">XOR</a></span>(java.util.Vector&lt;<a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">GenomicElement</a>&gt;&nbsp;v)</code>
<div class="block">returns GenomicElement that holds the regions covered by exactly one of the overlapping GenomicElements</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../../../at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/GenomicElement.html" title="class in at.ac.oeaw.cemm.bsf.vcffilter.genomeutils">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../../../../../../../../overview-tree.html">Tree</a></li>
<li><a href="../../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../../index.html?at/ac/oeaw/cemm/bsf/vcffilter/genomeutils/class-use/GenomicElement.html" target="_top">Frames</a></li>
<li><a href="GenomicElement.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
